src/ITensorNetworks/peps.jl

```diff
@@ -0,0 +1,142 @@
+"""
+A finite size PEPS type.
+"""
+mutable struct PEPS
+  data::Matrix{ITensor}
+end
+
+PEPS(Nx::Int, Ny::Int) = PEPS(Matrix{ITensor}(undef, Nx, Ny))
+
+"""
+    PEPS([::Type{ElT} = Float64, sites; linkdims=1)
+Construct an PEPS filled with Empty ITensors of type `ElT` from a collection of indices.
+Optionally specify the link dimension with the keyword argument `linkdims`, which by default is 1.
+"""
+function PEPS(::Type{T}, sites::Matrix{<:Index}; linkdims::Integer=1) where {T<:Number}
+  Ny, Nx = size(sites)
+  tensor_grid = Matrix{ITensor}(undef, Ny, Nx)
+  # we assume the PEPS at least has size (2,2). Can generalize if necessary
+  @assert(Nx >= 2 && Ny >= 2)
+
+  lh = Matrix{Index}(undef, Ny, Nx - 1)
+  for ii in 1:(Nx - 1)
+    for jj in 1:(Ny)
+      lh[jj, ii] = Index(linkdims, "Lh,$jj,$ii")
+    end
+  end
+  lv = Matrix{Index}(undef, Ny - 1, Nx)
+  for ii in 1:(Nx)
+    for jj in 1:(Ny - 1)
+      lv[jj, ii] = Index(linkdims, "Lv,$jj,$ii")
+    end
+  end
+
+  # boundary cases
+  tensor_grid[1, 1] = emptyITensor(T, lh[1, 1], lv[1, 1], sites[1, 1])
+  tensor_grid[1, Nx] = emptyITensor(T, lh[1, Nx - 1], lv[1, Nx], sites[1, Nx])
+  tensor_grid[Ny, 1] = emptyITensor(T, lh[Ny, 1], lv[Ny - 1, 1], sites[Ny, 1])
+  tensor_grid[Ny, Nx] = emptyITensor(T, lh[Ny, Nx - 1], lv[Ny - 1, Nx], sites[Ny, Nx])
+  for ii in 2:(Nx - 1)
+    tensor_grid[1, ii] = emptyITensor(T, lh[1, ii], lh[1, ii - 1], lv[1, ii], sites[1, ii])
+    tensor_grid[Ny, ii] = emptyITensor(
+      T, lh[Ny, ii], lh[Ny, ii - 1], lv[Ny - 1, ii], sites[Ny, ii]
+    )
+  end
+
+  # inner sites
+  for jj in 2:(Ny - 1)
+    tensor_grid[jj, 1] = emptyITensor(T, lh[jj, 1], lv[jj, 1], lv[jj - 1, 1], sites[jj, 1])
+    tensor_grid[jj, Nx] = emptyITensor(
+      T, lh[jj, Nx - 1], lv[jj, Nx], lv[jj - 1, Nx], sites[jj, Nx]
+    )
+    for ii in 2:(Nx - 1)
+      tensor_grid[jj, ii] = emptyITensor(
+        T, lh[jj, ii], lh[jj, ii - 1], lv[jj, ii], lv[jj - 1, ii], sites[jj, ii]
+      )
+    end
+  end
+
+  return PEPS(tensor_grid)
+end
+
+PEPS(sites::Matrix{<:Index}, args...; kwargs...) = PEPS(Float64, sites, args...; kwargs...)
+
+PEPS(data::Array, size::Tuple{Int64,Int64}) = PEPS(reshape(data, size[1], size[2]))
+
+function randomizePEPS!(P::PEPS)
+  dimy, dimx = size(P.data)
+  for ii in 1:dimx
+    for jj in 1:dimy
+      randn!(P.data[jj, ii])
+      normalize!(P.data[jj, ii])
+    end
+  end
+end
+
+function Base.:+(A::PEPS, B::PEPS)
```

Comment:
I think we should pick a different name for this functionality.

I think of `A::PEPS + B::PEPS` as adding two quantum states `|A> + |B>` where the states are in the form of a PEPS, which is not the same as adding the individual tensors at each site (it is a more complicated operation which would result in a PEPS `|C>` with a generally larger bond dimension).

I would consider this operation to be more like a map or broadcast, so in Julia notation would be `map(*, A::PEPS, B::PEPS)`, `broadcast(+, A::PEPS, B::PEPS)`, or `A .+ B`. To use the last notation, we would need to define broadcasting for the PEPS type, which is slightly involved (https://docs.julialang.org/en/v1/manual/interfaces/#man-interfaces-broadcasting).

For now, I would suggest just defining this with a different name, so something like:
```julia
function map_add(A::PEPS, B::PEPS)
  return PEPS(map(+, A.data, B.data))
end
```

Reply:
Done

Reply:
Note that currently the Base.:+ definition for PEPS is needed to get Zygote working.

One situation for this is to add the gradient d(inner)/d(peps) + d(inner)/d(peps_prime) * d(peps_prime)/d(peps). In this case the addition is automatically used to sum the gradients from different parts.

Reply:
That is a bit unfortunate, I made an issue about it here: #17

Maybe there is a different function besides `+` which was can overload for the sake of Zygote, or we can raise an issue about it with Zygote.